hal/halx86/up/pic.S

```diff
@@ -45,4 +48,66 @@ CallIntHandler:
     jmp eax
 .ENDP
 
+PUBLIC _HalpDispatchInterrupt
+.PROC _HalpDispatchInterrupt
+    FPO 0, 0, 0, 0, 0, FRAME_FPO
+
+    /* Call the C function */
+    call _HalpDispatchInterrupt2
+
+    /* Check if we got a pointer back */
+    test eax, eax
+    jnz CallIntHandler2
+
+    /* No? Just return */
+    ret
+
+CallIntHandler2:
+    /* Optimize the tail call to avoid stack overflow */
+    jmp eax
+.ENDP
+
+PUBLIC _HalEndSystemInterrupt@8
+.PROC _HalEndSystemInterrupt@8
+    FPO 0, 2, 0, 0, 0, FRAME_FPO
+
+    /* Call the C function with the same arguments we got */
+    push [esp+8]
+    push [esp+8]
```

Comment:
In terms of readability, that hurts.
But I'm not sure how it could be made "more readable".

Reply:
We control the inner function so we could make it fastcall. That would change this to:

```
mov ecx, [esp+4]
mov edx, [esp+8]
```

Better? :p

Reply:
Ah yes! :-)